analyzer/metadata-plugin/src/main/java/eu/fasten/analyzer/metadataplugin/MetadataDatabasePlugin.java

```diff
@@ -0,0 +1,244 @@
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package eu.fasten.analyzer.metadataplugin;
+
+import eu.fasten.analyzer.metadataplugin.db.MetadataDao;
+import eu.fasten.analyzer.metadataplugin.db.PostgresConnector;
+import eu.fasten.core.data.ExtendedRevisionCallGraph;
+import eu.fasten.core.plugins.KafkaConsumer;
+import java.io.IOException;
+import java.sql.SQLException;
+import java.sql.Timestamp;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import org.apache.kafka.clients.consumer.ConsumerRecord;
+import org.jooq.DSLContext;
+import org.jooq.exception.DataAccessException;
+import org.jooq.impl.DSL;
+import org.json.JSONException;
+import org.json.JSONObject;
+import org.pf4j.Extension;
+import org.pf4j.Plugin;
+import org.pf4j.PluginWrapper;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class MetadataDatabasePlugin extends Plugin {
+
+    public MetadataDatabasePlugin(PluginWrapper wrapper) {
+        super(wrapper);
+    }
+
+    @Extension
+    public static class MetadataPlugin implements KafkaConsumer<String> {
+
+        private final String topic = "opal_callgraphs";
+        private final int transactionRestartLimit = 3;
+
+        private DSLContext dslContext;
+        private boolean processedRecord = false;
+        private String pluginError = "";
+        private final Logger logger = LoggerFactory.getLogger(MetadataPlugin.class.getName());
+        private boolean restartTransaction = false;
+
+        public MetadataPlugin() throws IOException, SQLException, IllegalArgumentException {
+            this(PostgresConnector.getDSLContext());
+        }
+
+        public MetadataPlugin(DSLContext dslContext) {
+            super();
+            this.dslContext = dslContext;
+        }
+
+        @Override
+        public List<String> consumerTopics() {
+            return new ArrayList<>(Collections.singletonList(topic));
+        }
+
+        @Override
+        public void consume(String topic, ConsumerRecord<String, String> record) {
+            var consumedJson = new JSONObject(record.value());
+            this.processedRecord = false;
+            this.restartTransaction = false;
+            this.pluginError = "";
+            int transactionRestartCount = 0;
+            do {
+                try {
+                    var callgraph = new ExtendedRevisionCallGraph(consumedJson);
```

Comment:
I separated error handling for transactions and creating ERCG. However, I don't really see a point in having `continue` when catching transaction exception since it would be a last statement in the loop anyway. Maybe I just misunderstood you. Could you elaborate that part please @gousiosg ?

Reply:
You should continue to the next input object after you deal with the `JSONException`.

Reply:
I extracted that part of the code (where I convert JSONObject into ERCG) out of the loop. So when it catches JSONException, it deals with it and returns from the method to consume the next record. Is this fine?

Reply:
Yes!